### DISCUSSION TITLE
Disable OTP Configuration in Flow

### DISCUSSION THREAD

**Discussion by BiasF** [keycloak/keycloak#21636] [Q&A]
Hi,

we have to ways for Auth in our Flow:

- Smartcard (Prefered way and most used)
- Username + Password + OTP

The problem is, most user are using the smartcard for login and will never use OTP and they will never configure a OTP.
These account will be open for phishing attacks or lost credentials (a person can steal the cerentials and is able to create a OTP for the unconfigured account).

What I want to do is to disable the configuration of the OTP inside the flow, the user should manually go to the accound console and configure the OTP there (The account Console has a flow which allows only Smartcard login) if he need a OTP login.
If a person tries to login with username + password and the OTP is not configured, the login should fail.

**Comment by ADASNET** [comment]
Hey,

I recently had to do something very similar, almost identical really with some small differences.
I added a new provider for the authentication flow and put it before the OTP form in the authentication flow.

Here it checks whether OTP is actually needed for the login to succeed, and if it is but they don't have it? fail the flow right there and then.
If it is and they do have it then send them on, same if they don't have at and it isn't needed.

Wish I could show you my implementation but I do not think I am allowed to do so.
Hope it helps all the same.

**Reply by BiasF** [reply]
Hi,
Thanks for your answer.
With which execution it's possible to check if OTP is already configured?

**Reply by ADASNET** [reply]
If you make you own authentication provider you will have acces to the AuthenticationFlowContext.
You can use this to get the session in which you can find the stored credentials for the user currently on the way through the flow.

The credentials will contain "otp" if its already configured.

**Reply by ADASNET** [reply]
If you need an example of how to set up your own authenticator provider Keycloak provides examples in the source code here:

![image](https://github.com/keycloak/keycloak/assets/106650650/78f90615-ef6e-4206-9d0b-5faeeae852e7)


And you will need to set up the authenticator and the authenticatorFactory, as well as associated file under the resources folder

**Comment by BiasF** [comment]
Hi,

thanks, but thats not possible in my environment :(
I have access to the Admin-Console only, because it's a hosted instance from our central IT Departement.

**Reply by ADASNET** [reply]
Sorry then I am unsure how to do it, I'd suggest contacting whoever is maintaining the solution and ask them to add this, otherwise I do not think there is a good way with out-the-box keycloak to do what you are looking to do.
